File: backend/app/routers/prompts.py
"""提示词管理 API

提供提示词的 CRUD 操作接口。
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_session
from app.prompts.registry import PromptRegistry
from app.prompts.schemas import (
    PromptCreate,
    PromptListResponse,
    PromptResetResponse,
    PromptResponse,
    PromptUpdate,
)

router = APIRouter(prefix="/prompts", tags=["prompts"])


@router.get("", response_model=PromptListResponse)
async def list_prompts(
    category: str | None = Query(None, description="过滤分类：agent/memory/skill/crawler"),
    include_inactive: bool = Query(False, description="是否包含禁用的提示词"),
    session: AsyncSession = Depends(get_session),
):
    """获取所有提示词列表

    返回所有提示词，包括默认值和自定义值。
    自定义值会标记 source=custom 并附带 default_content。
    """
    registry = PromptRegistry(session)
    items = await registry.list_all(category=category, include_inactive=include_inactive)
    return PromptListResponse(items=items, total=len(items))


@router.get("/{key:path}", response_model=PromptResponse)
async def get_prompt(
    key: str,
    session: AsyncSession = Depends(get_session),
):
    """获取单个提示词

    优先返回数据库中的自定义值，无则返回默认值。
    """
    registry = PromptRegistry(session)
    prompt = await registry.get(key)
    if not prompt:
        raise HTTPException(status_code=404, detail=f"提示词 {key} 不存在")
    return prompt


@router.put("/{key:path}", response_model=PromptResponse)
async def update_prompt(
    key: str,
    data: PromptUpdate,
    session: AsyncSession = Depends(get_session),
):
    """更新提示词

    如果数据库无记录，会基于默认值创建新的自定义记录。
    """
    registry = PromptRegistry(session)
    try:
        prompt = await registry.update(key, data)
        await session.commit()
        return prompt
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("", response_model=PromptResponse)
async def create_prompt(
    data: PromptCreate,
    session: AsyncSession = Depends(get_session),
):
    """创建新的自定义提示词

    用于创建不在默认值中的全新提示词。
    """
    registry = PromptRegistry(session)
    try:
        prompt = await registry.create(data)
        await session.commit()
        return prompt
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{key:path}/reset", response_model=PromptResetResponse)
async def reset_prompt(
    key: str,
    session: AsyncSession = Depends(get_session),
):
    """重置提示词为默认值

    删除数据库中的自定义记录，恢复使用默认值。
    """
    registry = PromptRegistry(session)
    try:
        prompt = await registry.reset(key)
        await session.commit()
        return PromptResetResponse(
            key=key,
            message="已重置为默认值",
            content=prompt.content,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{key:path}")
async def delete_prompt(
    key: str,
    session: AsyncSession = Depends(get_session),
):
    """删除自定义提示词

    仅能删除非默认提示词。有默认值的提示词请使用 reset。
    """
    registry = PromptRegistry(session)
    try:
        deleted = await registry.delete(key)
        if not deleted:
            raise HTTPException(status_code=404, detail=f"提示词 {key} 不存在")
        await session.commit()
        return {"message": f"已删除提示词 {key}"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/lib/api/prompts.ts
/**
 * 提示词管理 API 客户端
 */

import { apiRequest } from "@/lib/api/client";

// ========== 类型定义 ==========

export type PromptCategory = "agent" | "memory" | "skill" | "crawler";
export type PromptSource = "default" | "custom";

export interface Prompt {
  key: string;
  category: PromptCategory;
  name: string;
  description: string | null;
  content: string;
  variables: string[];
  source: PromptSource;
  is_active: boolean;
  default_content: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface PromptListResponse {
  items: Prompt[];
  total: number;
}

export interface PromptUpdate {
  name?: string;
  description?: string;
  content?: string;
  is_active?: boolean;
}

export interface PromptCreate {
  key: string;
  category: PromptCategory;
  name: string;
  description?: string;
  content: string;
  variables?: string[];
}

export interface PromptResetResponse {
  key: string;
  message: string;
  content: string;
}

// ========== API 函数 ==========

const BASE_URL = "/api/v1/admin/prompts";

/**
 * 获取提示词列表
 */
export async function listPrompts(params?: {
  category?: PromptCategory;
  include_inactive?: boolean;
}): Promise<PromptListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.category) searchParams.set("category", params.category);
  if (params?.include_inactive) searchParams.set("include_inactive", "true");

  const query = searchParams.toString();
  const url = query ? `${BASE_URL}?${query}` : BASE_URL;
  return apiRequest<PromptListResponse>(url);
}

/**
 * 获取单个提示词
 */
export async function getPrompt(key: string): Promise<Prompt> {
  return apiRequest<Prompt>(`${BASE_URL}/${encodeURIComponent(key)}`);
}

/**
 * 更新提示词
 */
export async function updatePrompt(
  key: string,
  data: PromptUpdate
): Promise<Prompt> {
  return apiRequest<Prompt>(`${BASE_URL}/${encodeURIComponent(key)}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 创建自定义提示词
 */
export async function createPrompt(data: PromptCreate): Promise<Prompt> {
  return apiRequest<Prompt>(BASE_URL, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * 重置提示词为默认值
 */
export async function resetPrompt(key: string): Promise<PromptResetResponse> {
  return apiRequest<PromptResetResponse>(
    `${BASE_URL}/${encodeURIComponent(key)}/reset`,
    {
      method: "POST",
    }
  );
}

/**
 * 删除自定义提示词
 */
export async function deletePrompt(key: string): Promise<void> {
  await apiRequest<void>(`${BASE_URL}/${encodeURIComponent(key)}`, {
    method: "DELETE",
  });
}

// ========== 辅助函数 ==========

export const PROMPT_CATEGORY_LABELS: Record<PromptCategory, string> = {
  agent: "Agent 提示词",
  memory: "记忆系统",
  skill: "技能生成",
  crawler: "爬虫提取",
};

export const PROMPT_CATEGORY_COLORS: Record<PromptCategory, string> = {
  agent: "bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300",
  memory: "bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-300",
  skill: "bg-purple-100 text-purple-800 dark:bg-purple-900 dark:text-purple-300",
  crawler: "bg-orange-100 text-orange-800 dark:bg-orange-900 dark:text-orange-300",
};

export const PROMPT_SOURCE_LABELS: Record<PromptSource, string> = {
  default: "默认",
  custom: "已自定义",
};
